### PR TITLE
Add UI Materials

### LINKED ISSUE #5607: Add support for applying a shader to a UI Image
## What problem does this solve or what need does it fill?

Being able to apply shaders to UI Images could open a lot of possibilities for nice UIs! Because as of right now I feel like we're a bit limited when we want to make animated UI components.

## What solution would you like?

With the current way to add a shader to an Image, being able to insert a `MaterialMesh2dBundle` to a UI could work!

## What alternative(s) have you considered?

Not sure about this one, but if there was a way to link a fragment shader to an Image directly, there would maybe be no need to change anything for the UI Images?

### PR DESCRIPTION
**Pull request by MarkusTheOrt**
# Objective

- Add Ui Materials so that UI can render more complex and animated widgets.
- Fixes #5607 

## Solution
- Create a UiMaterial trait for specifying a Shader Asset and Bind Group Layout/Data.
- Create a pipeline for rendering these Materials inside the Ui layout/tree.
- Create a MaterialNodeBundle for simple spawning.

## Changelog

- Created a `UiMaterial` trait for specifying a Shader asset and Bind Group.
- Created a `UiMaterialPipeline` for rendering said Materials.
- Added Example [`ui_material` ](https://github.com/MarkusTheOrt/bevy/blob/ui_material/examples/ui/ui_material.rs) for example usage.
- Created [`UiVertexOutput`](https://github.com/MarkusTheOrt/bevy/blob/ui_material/crates/bevy_ui/src/render/ui_vertex_output.wgsl) export as VertexData for shaders.
- Created [`material_ui`](https://github.com/MarkusTheOrt/bevy/blob/ui_material/crates/bevy_ui/src/render/ui_material.wgsl) shader as default for both Vertex and Fragment shaders.

### CODE REVIEW

**Comment by IceSentry** [comment]
Could you add a simple example showing how to use this? I can figure it out easily but I believe this feature deserve a standalone example.

**Comment by MarkusTheOrt** [comment]
I just added a simple example showcasing the feature. 
I am thinking about ditching the use of `UiPipeline` inside the `UiMaterialPipeline` so that I can get rid of the default image bindings. Not every shader requires a texture, and if one is required it can be passed using the material.

Next up I do some renaming and documenting

**Comment by MarkusTheOrt** [comment]
I did now remove the usage of `UiPipeline`, and copied the relevant code into the `UiMaterialPipeline`.
I also did split up the additions, with essentially only the `UiMaterial` trait and its accompanying `UiMaterialKey` being inside the `material.rs` and the rest being in `render/material_pipeline.rs`.

File: crates/bevy_ui/src/node_bundles.rs [hunk 1]
```diff
@@ -326,3 +326,49 @@ impl Default for ButtonBundle {
         }
     }
 }
+
+/// A Ui Node that renders a shader
```
Comment (ickshonpe) [inline crates/bevy_ui/src/node_bundles.rs:330]: Seems back to front, the node doesn't render the shader.
```suggestion
/// A UI node that is rendered using a [`UiMaterial`]
```

File: crates/bevy_ui/src/node_bundles.rs [hunk 2]
```diff
@@ -326,3 +326,49 @@ impl Default for ButtonBundle {
...
+#[derive(Bundle, Clone, Debug)]
+pub struct MaterialNodeBundle<M: UiMaterial> {
+    /// Describes the logical size of the node
+    pub node: Node,
+    /// Styles which control the layout (size and position) of the node and it's children
+    /// In some cases these styles also affect how the node drawn/painted.
+    pub style: Style,
+    /// The Material of the Node.
```
Comment (ickshonpe) [inline crates/bevy_ui/src/node_bundles.rs:338]:
```suggestion
    /// The [`UiMaterial`] used to render the node.
```

File: examples/ui/ui_material.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+//! This example illustrates how to create a Node that shows a Material and driver parameters from
```
Comment (ickshonpe) [inline examples/ui/ui_material.rs:1]:
```suggestion
//! This example illustrates how to create a Node that shows a Material and derives parameters from
```

"driver" meant to be "derives"?

File: crates/bevy_ui/src/render/material_pipeline.rs [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+
+    fn finish(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app.init_resource::<UiMaterialPipeline<M>>();
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct UiMatMeta<M: UiMaterial> {
```
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:103]:
```suggestion
pub struct UiMaterialMeta<M: UiMaterial> {
```

Intuitively, `Mat` implies something to do with matrices. It's obviously not once you look at the type but could be confusing at a glance.

File: crates/bevy_ui/src/render/material_pipeline.rs [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+        Self {
+            vertices: BufferVec::new(BufferUsages::VERTEX),
+            view_bind_group: Default::default(),
+            marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Component)]
+pub struct MatUiBatch<M: UiMaterial> {
```
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:120]: Same with this, I'd go with:
```suggestion
pub struct UiMaterialBatch<M: UiMaterial> {
```

File: crates/bevy_ui/src/render/material_pipeline.rs [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+        pass.set_bind_group(
+            I,
+            ui_meta.into_inner().view_bind_group.as_ref().unwrap(),
+            &[view_uniform.offset],
+        );
+        RenderCommandResult::Success
+    }
+}
+
+pub struct DrawUiMatNode<M>(PhantomData<M>);
```
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:280]:
```suggestion
pub struct DrawUiMaterialNode<M>(PhantomData<M>);
```

File: crates/bevy_ui/src/render/material_pipeline.rs [hunk 7]
```diff
@@ -0,0 +1,659 @@
...
+            ),
+        ]
+        .map(|pos| pos / extent);
+
+        for i in QUAD_INDICES {
+            ui_meta.vertices.push(crate::UiVertex {
+                position: positions_clipped[i].into(),
+                uv: uvs[i].into(),
+                color: Color::WHITE.into(),
+                mode: 1,
```
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:462]: If `mode` is always going to be set to `0`, shouldn't we add a separate `UiMaterialVertex` with `mode` removed? The batching optimization it allows can't be implemented with this design anyway (and probably aren't necessary).

Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/material_pipeline.rs]: I did just as you suggested, had to make some other changes as well 👍 

thanks for the suggestion!

File: examples/ui/ui_material.rs (hunk 3 above)
Comment (MarkusTheOrt) [reply examples/ui/ui_material.rs:1]: I wanted to write "drives" as in "one can use a system that drives the parameters of this material".

I don't think its a great description, but can't seem to find the words for a better one right now.

File: crates/bevy_ui/src/node_bundles.rs [hunk 8]
```diff
@@ -326,3 +326,49 @@ impl Default for ButtonBundle {
         }
     }
 }
+
```
[the remark this answers is not in the document]
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/node_bundles.rs:329]: Great suggestion, I agree with you there, I wasn't aware of `ContentSize`

I'd like to get this PR done and would be happy to take on the `MaterialImageBundle` as a followup once it is ready

**Comment by MarkusTheOrt** [comment]
I did Change some of the names like @ickshonpe suggested.

I am thinking that it would be a good addition to add the [bevy_render::globals](https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/globals.rs) into the default shader bindings.
I believe it is a sensible default since `time` can be useful when doing anything animated within a shader.

File: assets/shaders/circle_shader.wgsl [hunk 9]
```diff
@@ -0,0 +1,48 @@
+// This shader draws a circular progress bar
+#import bevy_render::view View
+
+struct VertexOutput {
```
Comment (st0rmbtw) [inline assets/shaders/circle_shader.wgsl:4]: Maybe move the VertexOutput into an import to make less boilerplate?

Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl:4]: I think this is the right approach. Should I just add the export at the top of the current shader, or should I create a new file that only houses the struct itself that has the export?

Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl:4]: @MarkusTheOrt Yep, that would be nice!

Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl]: I added the output inside the `material.wgsl` shader and have also implemented it in the `circle_shader.wgsl` 👍

Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl]: @MarkusTheOrt Oh, sorry, I misread your comment. I think that a new file `ui_vertex_output.wgsl` that contains only `UiVertexOutput` would be better, like it's done in https://github.com/bevyengine/bevy/blob/main/crates/bevy_sprite/src/mesh2d/mesh2d_vertex_output.wgsl

**Comment by ickshonpe** [comment]
> > The issue is that prepare_uinodes and prepare_uimaterial_nodes are working independently of each other.
> 
> I am not sure if there is anything that can be done here. Disabling all batching is sub-optimal to say the least. Maybe there can be batching for each _layer_ in the Ui?

Batching doesn't have to be disabled, that was just a quick hack to diagnose the problem. There are a couple of approaches that could work, maybe a shared list resource that keeps track of what has been extracted at each stack index so the prepare systems know when to break up the batches. Or just one prepare system that handles batching for all the different extracted items.

File: assets/shaders/circle_shader.wgsl (hunk 9 above)
Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl]: Thank you! :)

**Comment by MarkusTheOrt** [comment]
> Batching doesn't have to be disabled, that was just a quick hack. There are a couple of approaches that could work, maybe a shared list resource that keeps track of what has been extracted at each stack index so the prepare systems know when to break up the batches. Or just one prepare system that handles batching for all the different extracted items.

I've been thinking about this, maybe a marker component for Material Nodes that can be queried for in `extract_uinodes` and then used in `prepare_uinodes` can be used.

**Comment by ickshonpe** [comment]
> EDIT: I've tried doing that and it seems to be working from a batching standpoint, looking at it in renderdoc it does two draw calls, one for the black square, borders, and the vertical bar, and one for the red bar, the circle still draws on top (I guess that is what your PR (#9062 ) fixes?

Sounds like all you need for it to work is to set the `z` field of each `UiBatch` and `UiMaterialBatch` equal to the stack index (`as f32`) of the last extracted UI node belonging to each batch, except for maybe some local ordering issues for nodes with more than one extracted item.

**Comment by MarkusTheOrt** [comment]
> Yes, sounds like you are almost there. All you should need to do is set the `z` field of each `UiBatch` and `UiMaterialBatch` equal to the stack index (`as f32`) of the last extracted UI node belonging to each batch. Maybe there might be some local ordering issues for nodes with more than one extracted item but that's not a big deal.

Got it working through that, thank you, I guess i'll make it clear in the initial post that this is blocked until the above mentioned PR is being merged 👍 and then do some final review myself before marking it ready ❤️ thanks for all the feedback and help so far!
Screenshot

**Comment by MarkusTheOrt** [comment]
Okay, I did look over the code quite a bit now and I believe this is ready to merge, just waiting for some feedback and reviews now! 👍

File: crates/bevy_ui/src/render/material_pipeline.rs [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+    mut pipelines: ResMut<SpecializedRenderPipelines<UiMaterialPipeline<M>>>,
+    pipeline_cache: Res<PipelineCache>,
+    render_materials: Res<RenderUiMaterials<M>>,
+    mut views: Query<(&ExtractedView, &mut RenderPhase<TransparentUi>)>,
+) where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    let draw_function = draw_functions.read().id::<DrawUiMaterial<M>>();
+
+    for (entity, extraced_uinode) in extracted_uinodes.uinodes.iter() {
```
Comment (MarkusTheOrt) [inline crates/bevy_ui/src/render/material_pipeline.rs:719]: Should be `extracted_uinode`

File: crates/bevy_ui/src/material.rs [hunk 11]
```diff
@@ -0,0 +1,129 @@
+use std::hash::Hash;
+
+use bevy_asset::Asset;
+use bevy_render::render_resource::{AsBindGroup, RenderPipelineDescriptor, ShaderRef};
+/// Materials are used alongside [`UiMaterialPlugin`](crate::UiMaterialPipeline) and [`MaterialNodeBundle`](crate::prelude::MaterialNodeBundle)
```
Comment (IceSentry) [inline crates/bevy_ui/src/material.rs:5]:
```suggestion

/// Materials are used alongside [`UiMaterialPlugin`](crate::UiMaterialPipeline) and [`MaterialNodeBundle`](crate::prelude::MaterialNodeBundle)
```

**Review by IceSentry — APPROVED** [review]
This seems really good to me. 

One thing though. Could you rename material and material_pipeline to ui_material and ui_material_pipeline (or alternatively ui_material_plugin). It will make it easier to navigate to the file from an editor. material_pipeline is a bit too generic in my opinion. 

It would also be good to document in that the view binding is present probably somewhere in the UitMaterial doc.

I think there should be a follow up PR that adds at the very least the `globals` binding but maybe something like mouse position too. It doesn't need to be in this PR though. It's just that I'm sure people will use this to make a shadertoy in bevy since it would be the easiest way to do that.

Despite the couple of nitpicks I mentioned I think this could be merged and improved upon after so I'll approve anyway.

File: crates/bevy_ui/src/render/material.wgsl [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+@vertex
+fn vertex(
+    @location(0) vertex_position: vec3<f32>,
+    @location(1) vertex_uv: vec2<f32>,
+    @location(2) vertex_color: vec4<f32>,
+) -> UiVertexOutput {
+    var out: UiVertexOutput;
+    out.uv = vertex_uv;
+    out.position = view.view_proj * vec4<f32>(vertex_position, 1.0);
+    out.color = vertex_color;
```
[the remark this answers is not in the document]
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/material.wgsl]: Now that you mention it, I don't think so. I think its best to remove it, and if there is a use case where vertex color makes sense then we can still add it back with a PR, its not a big change.

File: assets/shaders/circle_shader.wgsl [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+// How smooth the border of the gradient should be
+const gradient_ease: f32 = 25.0;
+// the width of the gradient
+const width = 0.25;
+const PI = 3.141592656;
+const TAU = 6.283185312;
+
+@fragment
+fn fragment(in: UiVertexOutput) -> @location(0) vec4<f32> {
```
[the remark this answers is not in the document]
Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl:12]: I agree, I've updated the example, the shader now only takes in one property (color) and just draws a simple circle with the color shift.

**Comment by ickshonpe** [comment]
Not forgotten about this, I'll try and do a review a bit later today after I've finished what I'm currently working on.

**Review by alice-i-cecile — APPROVED** [review]
Straightforward and appears to be consistent with how our other materials work. The code quality and docs are good (or at least, on par with the rest of rendering) and this is a very useful feature.

**Comment by nicopap** [comment]
@superdump You were complaining that we were duplicating the material code between 2d and 3d, this would add another instance of such duplication.

**Comment by superdump** [comment]
Indeed. I really want to consolidate 2D and 3D materials and minimise the differences else developing any improvements requires implementing the same thing multiple times.

File: Cargo.toml [hunk 14]
```diff
@@ -2319,6 +2319,17 @@ description = "Demonstrates resizing and responding to resizing a window"
...
 wasm = true
 
+[[example]]
+name = "ui_material"
+path = "examples/ui/ui_material.rs"
+doc-scrape-examples = true
+
+[package.metadata.example.ui_material]
+name = "UI Material"
+description = "Demonstrates Creating and using custom Ui materials"
```
Comment (mockersf) [inline Cargo.toml:2329]:
```suggestion
description = "Demonstrates creating and using custom Ui materials"
```

File: crates/bevy_ui/src/render/mod.rs [hunk 15]
```diff
@@ -622,14 +624,14 @@ impl Default for UiMeta {
     }
 }
 
-const QUAD_VERTEX_POSITIONS: [Vec3; 4] = [
+pub const QUAD_VERTEX_POSITIONS: [Vec3; 4] = [
```
Comment (mockersf) [inline crates/bevy_ui/src/render/mod.rs:627]:
```suggestion
pub(crate) const QUAD_VERTEX_POSITIONS: [Vec3; 4] = [
```
or document it

File: crates/bevy_ui/src/render/mod.rs [hunk 16]
```diff
@@ -622,14 +624,14 @@ impl Default for UiMeta {
...
     Vec3::new(-0.5, -0.5, 0.0),
     Vec3::new(0.5, -0.5, 0.0),
     Vec3::new(0.5, 0.5, 0.0),
     Vec3::new(-0.5, 0.5, 0.0),
 ];
 
-const QUAD_INDICES: [usize; 6] = [0, 2, 3, 0, 1, 2];
+pub const QUAD_INDICES: [usize; 6] = [0, 2, 3, 0, 1, 2];
```
Comment (mockersf) [inline crates/bevy_ui/src/render/mod.rs:634]:
```suggestion
pub(crate) const QUAD_INDICES: [usize; 6] = [0, 2, 3, 0, 1, 2];
```
or document it

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 17]
```diff
@@ -0,0 +1,737 @@
...
+
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct UiMaterialVertex {
+    pub position: [f32; 3],
+    pub uv: [f32; 2],
+}
+
+/// in this [`UiMaterialPipeline`] there is (currently) no batching going on.
+/// Therefore the [`UiMaterialBatch`] is more akin to a draw call.
```
Comment (mockersf) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:142]:
```suggestion
// in this [`UiMaterialPipeline`] there is (currently) no batching going on.
// Therefore the [`UiMaterialBatch`] is more akin to a draw call.
```
This seems to be more a code comment than a doc

**Comment by mockersf** [comment]
I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them

**Comment by MarkusTheOrt** [comment]
> I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them

I don't quite understand what you mean.
Don't borders get extracted and rendered just as colored ui nodes?

File: crates/bevy_ui/src/render/ui.wgsl [hunk 18]
```diff
@@ -39,4 +39,4 @@ fn fragment(in: VertexOutput) -> @location(0) vec4<f32> {
         color = in.color;
     }
     return color;
-}
+}
```
Comment (superdump) [inline crates/bevy_ui/src/render/ui.wgsl:42]: This change should be removed. :)

Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui.wgsl]: I might have made a complete fool out of myself here :P 
but is this a windows line ending or was that just a joke that completely went past myself?

**Comment by alice-i-cecile** [comment]
No longer controversial, but [waiting on review from @superdump](https://discord.com/channels/691052431525675048/1161381998925991986/1165065824030179448).

**Comment by mockersf** [comment]
> > I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them
> 
> I don't quite understand what you mean. Don't borders get extracted and rendered just as colored ui nodes?

Yes, if you add a `BorderColor` component.

With this PR, the material can be drawn over the border, but we don't know that as it's given the size of the node, border included. I would like to see the border width on each side as part of the data passed to the shader so that the material can decide how to render its borders

**Comment by MarkusTheOrt** [comment]
Okay, I have added the border sizes (in UV _size_ if it can be called that) 
so if you have a 100px square with a 10px border the values passed will be [0.1, 0.1, 0.1, 0.1] in the order of [left, right, top. bottom]

File: crates/bevy_ui/src/node_bundles.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (superdump) [reply crates/bevy_ui/src/node_bundles.rs:345]: Couldn’t the image just be a member of the material type? @ickshonpe

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 19]
```diff
@@ -0,0 +1,779 @@
...
+use bevy_utils::{FloatOrd, HashMap, HashSet};
+use bevy_window::{PrimaryWindow, Window};
+use bytemuck::{Pod, Zeroable};
+
+use crate::{
+    resolve_border_thickness, CalculatedClip, Node, RenderUiSystem, Style, TransparentUi,
+    UiMaterial, UiMaterialKey, UiScale, UiStack, QUAD_INDICES, QUAD_VERTEX_POSITIONS,
+};
+
+pub const MATERIAL_UI_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
```
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:51]:
```suggestion
pub const UI_MATERIAL_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 20]
```diff
@@ -0,0 +1,779 @@
...
+
+    fn finish(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app.init_resource::<UiMaterialPipeline<M>>();
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct UiMaterialMeta<M: UiMaterial> {
```
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:119]: Does this need to be per material or would one for all materials + instance offsets be enough?

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 21]
```diff
@@ -0,0 +1,779 @@
...
+                topology: PrimitiveTopology::TriangleList,
+                strip_index_format: None,
+            },
+            depth_stencil: None,
+            multisample: MultisampleState {
+                count: 1,
+                mask: !0,
+                alpha_to_coverage_enabled: false,
+            },
+            label: Some("material_ui_pipeline".into()),
```
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:220]:
```suggestion
            label: Some("ui_material_pipeline".into()),
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs (hunk 20 above)
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs:119]: I don’t think so. Should I initialize this one in the normal zu initialization or is there another way to do it once?

**Comment by MarkusTheOrt** [comment]
I’ll resolve these as soon as possible (in about one hour or so)

File: crates/bevy_ui/src/render/ui_material_pipeline.rs (hunk 20 above)
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs]: Apologies for the obviously bad question, I did read up on it and I updated everything. Now uses only one UI Meta (like the normal UI does)

File: crates/bevy_ui/src/render/ui_vertex_output.wgsl [hunk 22]
```diff
@@ -0,0 +1,9 @@
+#define_import_path bevy_ui::ui_vertex_output
+
+// The Vertex output of the default vertex shader for the Ui Material pipeline.
+struct UiVertexOutput {
+    @location(0) uv: vec2<f32>,
+    // The size of the borders in UV space. order is Left, Right, Top, Bottom.
```
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_vertex_output.wgsl:6]:
```suggestion
    // The size of the borders in UV space. Order is Left, Right, Top, Bottom.
```

File: crates/bevy_ui/src/ui_material.rs [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+///         }),
+///         ..Default::default()
+///     });
+/// }
+/// ```
+/// In WGSL shaders, the material's binding would look like this:
+///
+/// If you only use the fragment shader make sure to import `UiVertexOutput` from
+/// `bevy_ui::ui_vertex_output` in your wgsl shader.
+/// Also note that binding group 0 is always taken with the view matrix.
```
Comment (JMS55) [inline crates/bevy_ui/src/ui_material.rs:70]:
```
/// Also note that binding group 0 is always taken with the view matrix.
```

This is a weird sentence, not sure what it's trying to say.

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 24]
```diff
@@ -0,0 +1,774 @@
...
+};
+use bevy_math::{Mat4, Rect, Vec2, Vec4Swizzles};
+use bevy_render::{
+    extract_component::ExtractComponentPlugin,
+    render_asset::RenderAssets,
+    render_phase::{
+        AddRenderCommand, DrawFunctions, PhaseItem, RenderCommand, RenderCommandResult,
+        RenderPhase, SetItemPipeline, TrackedRenderPass,
+    },
+    render_resource::{
```
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:25]: Replace with a * import, see https://github.com/bevyengine/bevy/pull/9847

File: crates/bevy_ui/src/render/ui_material_pipeline.rs [hunk 25]
```diff
@@ -0,0 +1,774 @@
...
+    resolve_border_thickness, CalculatedClip, Node, RenderUiSystem, Style, TransparentUi,
+    UiMaterial, UiMaterialKey, UiScale, UiStack, QUAD_INDICES, QUAD_VERTEX_POSITIONS,
+};
+
+pub const UI_MATERIAL_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
+
+const UI_VERTEX_OUTPUT_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10123618247720234751);
+
+/// Adds the necessary ECS resources and render logic to enable rendering entities using the given
+/// [`UiMaterial`] asset type (which includes [`UiMaterial`] types).
```
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:56]: The last part of this sentence is confusing to me.

**Review by JMS55 — APPROVED** [review]
Code looks generally fine, although I'm not familiar with any of the UI stuff. Some small doc nits.

File: crates/bevy_ui/src/ui_material.rs (hunk 23 above)
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/ui_material.rs:70]: this is supposed to say that bind group 0 (in wgsl `@group(0)`) is always bound to the View matrix from `bevy_render`.

from the ui shader:
```wgsl
#import bevy_render::view::View

@group(0) @binding(0)
var<uniform> view: View;
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs (hunk 25 above)
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs:38]: I removed this part as it is confusing and doesn't convey any extra information. (I copied this from the Mesh2d Material pipeline. 👍

File: crates/bevy_ui/src/ui_material.rs (hunk 23 above)
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/ui_material.rs]: I changed the wording a bit and added a link to the View Uniform: 
`/// Also note that bind group 0 is always bound to the [`View Uniform`](bevy_render::view::ViewUniform).`

**Comment by alice-i-cecile** [comment]
I'm less sleep-deprived now and feeling comfortable re-reviewing this. I'm happy with the quality of code, the docs are great, and this is useful abstraction. Merging for 0.12 :)

**Pull request merged**